web_monitoring/db.py

```diff
@@ -1,47 +1,608 @@
+    res.raise_for_status()
+    result = res.json()
+    # In place, replace datetime strings with datetime objects.
+    for v in result['data']:
+        v['created_at'] = parse_timestamp(v['created_at'])
+        v['updated_at'] = parse_timestamp(v['updated_at'])
+        v['capture_time'] = parse_timestamp(v['capture_time'])
+    return result
+
+
+@ensure_credentials
+def list_versions(chunk=None, chunk_size=None,
+                  start_date=None, end_date=None,
+                  source_type=None, version_hash=None,
+                  source_metadata=None):
+    """
+    List all Versions, optionally filtered by serach criteria.
+
+    Parameters
+    ----------
+    chunk : integer, optional
+        pagination parameter
+    chunk_size : integer, optional
+        number of items per chunk
+    start_date : datetime, optional
+    end_date : datetime, optional
+    source_type : string, optional
+    version_hash : string, optional
+    source_metadata : dict, optional
+        Examples:
+
+        * ``{'version_id': 12345678}``
+        * ``{'account': 'versionista1', 'has_content': True}``
+
+    Returns
+    -------
+    response : dict
+    """
+    params = {'chunk': chunk,
+              'chunk_size': chunk_size,
+              'capture_time': time_range_string(start_date, end_date),
+              'source_type': source_type,
+              'version_hash': version_hash}
+    if source_metadata is not None:
+        for k, v in source_metadata.items():
+            params[f'source_metadata[{k}]'] = v
+    url = f'{api_url()}/versions'
+    res = requests.get(url, auth=auth(), params=params)
+    res.raise_for_status()
+    result = res.json()
+    # In place, replace datetime strings with datetime objects.
+    for v in result['data']:
+        v['created_at'] = parse_timestamp(v['created_at'])
+        v['updated_at'] = parse_timestamp(v['updated_at'])
+        v['capture_time'] = parse_timestamp(v['capture_time'])
+    return result
+
+
+@ensure_credentials
+def get_version(version_id):
+    """
+    Lookup a specific Version by ID.
+
+    Parameters
+    ----------
+    version_id : string
+
+    Returns
+    -------
+    response : dict
+    """
+    url = f'{api_url()}/versions/{version_id}'
+    res = requests.get(url, auth=auth())
+    res.raise_for_status()
+    result = res.json()
+    data = result['data']
+    data['capture_time'] = parse_timestamp(data['capture_time'])
+    data['updated_at'] = parse_timestamp(data['updated_at'])
+    data['created_at'] = parse_timestamp(data['created_at'])
+    return result
+
+
+@ensure_credentials
+def post_version(page_id, uuid, capture_time, uri, version_hash,
+                 source_type, source_metadata=None):
+    """
+    Submit one new Version.
+
+    See :func:`post_versions` for a more efficient bulk importer.
+
+    Parameters
+    ----------
+    page_id : string
+        Page to which the Version is associated
+    uuid : string
+        a new, unique Version ID (UUID4)
```

Comment:
Is this intentionally non-optional?

Reply:
Sure, we can generate one if `None`.

Reply:
I guess I was more noting that the API server will already do exactly that logic for you.